src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionList.java

```diff
@@ -275,7 +296,10 @@ public void onFailure(Throwable t) {
                     }
                 }
 
-            }, MoreExecutors.directExecutor()); // TODO: Unclear if we need to run this asynchronously or if StepExecution.onResume has any particular thread requirements.
+            }, STEP_RESUMPTION_EXECUTOR_SERVICE); // We always hold RunMap and WorkflowRun locks here, so we resume steps on a different thread to avoid potential deadlocks. See JENKINS-67351.
         }
     }
+
+    // Using an unbounded thread pool to avoid potential deadlock as in JENKINS-25890. TODO: Unclear if this is actually necessary.
```

Comment:
`Timer` is probably good enough? Not sure.

Reply:
Yeah, maybe I'll switch back to `Timer` for now and then if there is related deadlock we can introduce an `ExecutorService`. In this case we are at least guaranteed to release the `RunMap` lock as long as the callback does not use the current thread, so even if `Timer` is saturated and the task to resume the step executions is not scheduled quickly, I don't think anything else would be blocked waiting for it.